### ISSUE TITLE
julia: stale cache files resulting from setting mtime=1 with new julia packaging

### ISSUE THREAD

**Issue by colinxs** [NixOS/nixpkgs#137926]
### Describe the bug

From the [Julia manual](https://docs.julialang.org/en/v1/manual/modules/)
> For file dependencies, a change is determined by examining whether the modification time (mtime) of each file loaded by include or added explicitly by include_dependency is unchanged, or equal to the modification time truncated to the nearest second (to accommodate systems that can't copy mtime with sub-second accuracy)

This means the Julia compilation cache will be stale for packages residing in /nix/store as they have their mtime set to 1. 

@ninjin I know we've spoken about this a few times before. Unfortunately, the Julia slack recently switched to the free edition so we lost our conversation history. Previously there was [this patch](https://raw.githubusercontent.com/NixOS/nixpkgs/d51713378b62d32bd63466023308e6615cb5bedc/pkgs/development/compilers/julia/allow_nix_mtime.patch) which presumably addressed this bug (cannot currently test since the compiled-from-source version of Julia is broken which prompted the recent change to julia-bin). I think the folks over at julialang/julia would consider this a bug so I'll file an issue over there and link it. I don't see a way to fix it in nixpkgs since we switched to using the upstream precompiled binaries.

### Steps To Reproduce

I put together a minimal example at [colinxs/julia-stale-cache-bug](https://github.com/colinxs/julia-stale-cache-bug). You can run it with `nix run 'github:colinxs/julia-stale-cache-bug#run-example'`. You can see the details in the file 'example.sh' in that repo. What follows is an abbreviated example.

Using the following flake as an example:

./flake.nix:
```
{
  inputs.nixpkgs.url = "nixpkgs/nixpkgs-unstable";
  outputs = { self, nixpkgs }:
    let 
      pkgs = nixpkgs.legacyPackages.x86_64-linux;
    in {
      packages.x86_64-linux.example = pkgs.writeShellScriptBin "example" ''
        ${pkgs.julia_16-bin}/bin/julia --startup-file=no --project=${./Example} -e 'using Example; Example.greet()'
      '';
    };
}
```
./Example/src/Example.jl:
```julia
module Example
include("foo.jl")
end
```
./Example/src/foo.jl:
```julia
greet() = println("Hello!")
```

Now run it (clearing the cache first):
```
> rm -rf ~/.julia/compiled/v1.6/Example
> nix run .#example
Hello!
```

Now modify ./Example/src/foo.jl to be:
```julia
greet() = println("Goodbye!")
```

Run it again:
```
> nix run .#example
Hello!
```

This is incorrect. Now clear the cache and run it:
```
> rm -rf ~/.julia/compiled/v1.6/Example
> nix run .#example
Goodbye!
```

Curiously, the stale cache only occurs when files are included from Example.jl. For example, the following works correctly:

./Example/src/Example.jl:
```julia
module Example
greet() = println("Hello!")
end
```
```
> nix run .#example
Hello!
```
Now modify ./Example/src/Example.jl again:
```julia
module Example
greet() = println("Goodbye!")
end
```
```
> nix run .#example
Goodbye!
```

### Notify maintainers
@ninjin @romanofski 

### Metadata
```
 - system: `"x86_64-linux"`
 - host os: `Linux 5.11.0-7620-generic, Pop!_OS, 21.04`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.4pre20210707_02dd6bb`
 - channels(root): `"nixpkgs-21.11pre309670.253aecf69ed"`
 - nixpkgs: `/nix/store/svybjwf9ab3iqiz8zwwqwbqzksl3gx8v-source`
```

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
# a list of nixos modules affected by the problem
module:
```

**Comment by ninjin** [comment]
On Tue 14 Sep 2021, Colin Summers wrote:
> 
> @ninjin I know we've spoken about this a few times before. Unfortunately, the Julia slack recently switched to the free edition so we lost our conversation history. Previously there was [this patch](https://raw.githubusercontent.com/NixOS/nixpkgs/d51713378b62d32bd63466023308e6615cb5bedc/pkgs/development/compilers/julia/allow_nix_mtime.patch) which presumably addressed this bug (cannot currently test since the compiled-from-source version of Julia is broken which prompted the recent change to julia-bin). I think the folks over at julialang/julia would consider this a bug so I'll file an issue over there and link it. I don't see a way to fix it in nixpkgs since we switched to using the upstream precompiled binaries.

Yes, I recall this discussion over on the Julia Slack. This is indeed a bug, but one that we can remedy in Nixpkgs (although I certainly would prefer if Julia changed their code upstream not to rely on timestamps as that comes with its own set of additional problems). I am not a maintainer for the Nixpkgs Julia source builds, but for the source builds fixing the bug would be as easy as to [revert this commit](https://github.com/NixOS/nixpkgs/commit/4c97cf823f025cfee3f8c2fc028ca0c0ab3369f2) (on a largely unrelated note, if someone is willing to lift patches from [my Julia fork](https://git.sr.ht/~ninjin/julia-nix), you can build Julia from source using dependencies provided by BinaryBuilder to compensate for where we currently have issues building Julia from source).

Now, for the patched precompiled binaries (that I am a maintainer for), I am pretty certain as I said over on the Julia Slack that they can be fixed by incorporating the very same patch. All we should need to do is to rebuild Julia’s base image, which (while a bit verbose in terms of code) is not difficult and I have this rebuild working since this spring since it would be necessary for more involved patching such as getting `Pkg.jl` to play nice with NixOS. If someone wants to take a stab at this, poke me and I am happy to push this branch somewhere public. Not sure when I will get around to pushing this myself, but it is at least likely to happen before the 21.11 cut-off.

A few side notes. It is not true that “we switched to using the upstream precompiled binaries” (I mean, `julia` still points to the source builds…), the source and binary-based packages coexist and are not at odds with each other as they have different strengths and weaknesses from the point of view of maintainers and end users.

**Comment by colinxs** [comment]
Looks like this was fixed in julialang/julia#43090! :+1:
